### PR TITLE
Fix #2311: Possible to get multiple invite_send achievements

### LINKED ISSUE #2311: Possible to get multiple invite_send achievements

### PR DESCRIPTION
**Pull request by can-007**
When DB is outside the challenge time frame, there are problems, possible to get multiple invite_send achievements is one of them.

This PR could provide some suggestions for the case that no challenge is currently going on:
- protect the `hasEarnedBonus()` of achievements.ts from destructing properties of null;
- disallow to get the `invite_send `and `invite_contribute_same_session `  achievements;

@rileyjshaw , I test it on my local environment not the docker one, it could work.
We had brief discussion regarding what it should look like when there is no challenge currently, I only remembered a suggestion to disallow the enrollment.
@phirework , FYI.

I can think of the 4 situations above for now, any more?

### CODE REVIEW

**Comment by phirework** [comment]
The redirect bug has been fixed on master, so you should be able to test it locally if you merge master back in. Looking at code now.

**Comment by can-007** [comment]
Not completely fixed this one, the account is null, should check this one first.
Because the first time user ever visited the common-voice, there won't be any information about any user, so the account is null.

**Comment by phirework** [comment]
@mbransn Can you confirm whether it's desired behaviour for users to be prevented from enrolling if there's no ongoing challenge? Don't we want users to be able to pre-register for a challenge before it starts?

**Comment by rileyjshaw** [comment]
> Don't we want users to be able to pre-register for a challenge before it starts?

Can confirm that we definitely want this.

**Comment by can-007** [comment]
Oh, that is different from what I thought. OK.
But, what could these pre-enrolled user do? Can they win bonus or see the challenge board?

**Comment by rileyjshaw** [comment]
> But, what could these pre-enrolled user do? Can they win bonus or see the challenge board?

@can-007 no, they would just be signed up and waiting for the challenge. The only difference is that they would have a Challenge tab on the dashboard (which would show the pre-challenge state)

**Comment by can-007** [comment]
@rileyjshaw , I see now, the UI is controlled by the constants. Nice.

**Comment by mbransn** [comment]
+1 to everything Riley said.

**Comment by rileyjshaw** [comment]
@can-007 what is the status on this? Does it require updates based on the above conversation, or should I review the code as-is?

**Comment by can-007** [comment]
Please review the code as-is after a proper rest. I updated the branch based on the above conversation.

**Comment by can-007** [comment]
A little bit more detail: this could prevent achieving multiple `invite_send `and `invite_contribute_same_session `achievements when now no challenges are going on.

File: server/src/lib/model/achievements.ts [hunk 1]
```diff
@@ -73,9 +74,9 @@ const bonus_condition_sql = {
     GROUP BY user_clients.client_id
     `,
   // Arguments: [client_id, client_id, challenge]
-  // Check if user has already earned the bonus
+  // Check if user has already earned the bonus, and if challenge is on-going now
   invite_send: `
-    SELECT earn.id IS NULL AS win_bonus, ? AS bonus_winner
+    SELECT (earn.id IS NULL) AND (NOW() BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)) AS win_bonus, ? AS bonus_winner
```
Comment (rileyjshaw) [inline server/src/lib/model/achievements.ts]: This should still be achievable before the challenge. We have a pre-challenge state that includes the invite modal.

File: server/src/lib/model/achievements.ts [hunk 2]
```diff
@@ -84,9 +85,9 @@ const bonus_condition_sql = {
     WHERE challenges.url_token = ?
     `,
   // Arguments: [client_id, client_id, challenge]
-  // Check if user has already won the bonus
+  // Check if user has already won the bonus, and if challenge is on-going now
   invite_contribute_same_session: `
-    SELECT earn.id IS NULL AS win_bonus, ? AS bonus_winner
+    SELECT (earn.id IS NULL) AND (NOW() BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)) AS win_bonus, ? AS bonus_winner
```
Comment (rileyjshaw) [inline server/src/lib/model/achievements.ts]: Same here - this should be available pre-challenge.

File: server/src/lib/model/achievements.ts [hunk 3]
```diff
@@ -119,7 +120,8 @@ export const hasEarnedBonus = async (
   client_id: string,
   challenge: ChallengeToken
 ) => {
-  const [[{ earned }]] = await db.query(
+  // do not destructure to [[{earned}]], just in case null values returned, in scenarios like challenge=null
+  const [[res]] = await db.query(
```
Comment (rileyjshaw) [inline server/src/lib/model/achievements.ts]: This is fixed in https://github.com/mozilla/voice-web/pull/2373, can be safely removed.

Comment (phirework) [reply server/src/lib/model/achievements.ts]: omg I'm so annoyed that this comment was sitting in a pull request all along 😅

Comment (can-007) [reply server/src/lib/model/achievements.ts]: Yeah.

File: server/src/lib/model/achievements.ts (hunk 2 above)
Comment (can-007) [reply server/src/lib/model/achievements.ts]: Done

File: server/src/lib/model/achievements.ts (hunk 1 above)
Comment (can-007) [reply server/src/lib/model/achievements.ts]: Good point.

**Pull request closed without merging**